### PR TITLE
WIP - Optional portion -  2D framework implementation for review

### PR DESCRIPTION
**Pull request by Piinks**
WIP for discussion in [this design document](https://docs.google.com/document/d/1C2hAq-gMAIx4Cbym7EMhVAcPZ2hmNxlgijpYmgTpmUU/edit?usp=sharing)

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scroll_view.dart [hunk 1]
```diff
@@ -1938,3 +1938,165 @@ class GridView extends BoxScrollView {
...
+/// A widget that scrolls in both dimensions.
+///
+/// A two-way scrollable widget consist of three pieces:
+///
+///  1. A [TwoDimensionalScrollable] widget, which listens for various user
+///     gestures and implements the interaction design for scrolling.
+///  2. A [TwoDimensionalViewport] widget, which implements the visual design
+///     for scrolling by displaying only a portion
+///     of the widgets inside the scroll view.
+///  3. A [TwoDimensionalChildDelgate], which provides the children visible in
```
Comment (loic-sharma) [inline packages/flutter/lib/src/widgets/scroll_view.dart:1953]:
```suggestion
///  3. A [TwoDimensionalChildDelegate], which provides the children visible in
```

File: packages/flutter/lib/src/widgets/two_dimensional_viewport.dart [hunk 2]
```diff
@@ -0,0 +1,785 @@
...
+  /// For the given [Axis], the nearest [PrimaryScrollController] surrounding
+  /// the widget is attached to the scroll position of that Axis.
+  ///
+  /// Defaults to [Axis.vertical].
+  final Axis mainAxis;
+
+  /// Returns the child with the given indices.
+  ///
+  /// Should return null if asked to build a widget with a greater
+  /// index than exists. If this returns null, [estimatedChildCount]
```
Comment (loic-sharma) [inline packages/flutter/lib/src/widgets/two_dimensional_viewport.dart:35]: Does `estimatedChildCount` still exist? I couldn't find it on this change.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/two_dimensional_viewport.dart:35]: Oh! That is a bit of copy pasta. Thank you!

**Comment by Piinks** [comment]
Closing this to keep it off the triage queue, discussion and feedback is still welcome though.

**Pull request closed without merging**